src/devices/Ssd1306/Ssd1306.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Iot.Device.Ssd1306.Command;
+using System;
+using System.Device.I2c;
+
+namespace Iot.Device.Ssd1306
+{
+    /// <summary>
+    /// A single-chip CMOS OLED/PLED driver with controller for organic/polymer
+    /// light emitting diode dot-matrix graphic display system. 
+    /// </summary>
+    public class Ssd1306 : IDisposable
+    {
+        private I2cDevice _i2cDevice;
+
+        /// <summary>
+        /// Initializes new instance of Ssd1306 device that will communicate using I2C bus.
+        /// A single-chip CMOS OLED/PLED driver with controller for organic/polymer
+        /// light emitting diode dot-matrix graphic display system. 
+        /// </summary>
+        /// <param name="i2cDevice">>The I2C device used for communication.</param>
+        public Ssd1306(I2cDevice i2cDevice)
+        {
+            _i2cDevice = i2cDevice;
+        }
+
+        /// <summary>
+        /// Send a command to the display controller.
+        /// </summary>
+        /// <param name="command">The command to send to the display controller.</param>
+        public void SendCommand(ICommand command)
+        {
+            byte[] commandBytes = command.GetBytes();
+            byte[] writeBuffer = new byte[commandBytes.Length + 1];
+            commandBytes.CopyTo(writeBuffer, 1);  // Control byte is 0x00 for first byte.
+
+            // Be aware there is a Continuation Bit in the Control byte and can be used
+            // to state (logic LOW) if there is only data bytes to follow.
+            // This binding separates commands and data by using SendCommand and SendData.
+
+            _i2cDevice.Write(writeBuffer);
+        }
+
+        /// <summary>
+        /// Send data to the display controller.
+        /// </summary>
+        /// <param name="data">The data to send to the display controller.</param>
+        public void SendData(byte[] data)
+        {
+            if (data == null)
+            {
+                throw new ArgumentNullException(nameof(data), "The data is invalid.");
+            }
+
+            byte[] writeBuffer = new byte[data.Length + 1];
```

Comment:
If `data == empty array`, do we want to send a `0x40` control byte? Also, can we do a length check on data and avoid the unnecessary call to CopyTo?

Also, why aren't we doing a stackalloc for < 128 data.Length?

Reply:
> If `data == empty array`

I'm assuming you mean include `if (data == null || data.Length == 0)` or is there a better approach?

> Also, why aren't we doing a stackalloc for < 128 data.Length?

I've honestly never used stackalloc, so you'll have to educate me (if you don't mind).  How would you use this for a length of 128 or less?

Reply:
> I'm assuming you mean include `if (data == null || data.Length == 0)` or is there a better approach?

That works. But I don't know what the expected behavior should be if the caller returns an empty array. Should we still write a continuation byte to the device or is an empty array considered invalid?


> Would you use something similar to your comment [here](https://github.com/dotnet/iot/pull/188#discussion_r253183418), but use 128 instead?

Yes. 128, 256, some reasonably constant is fine. If you expect it to almost never be above 32, choose that instead.

```C#
const int StackThreshold = 256;

byte[] commandBytes = command.GetBytes();

if (commandBytes == null) throw; // TODO: appropriate exception

if (commandBytes.IsEmpty) return; // TODO: Do nothing? throw exception? something else?

Span<byte> buffer = commandBytes.Length < StackThreshold ?
   stackalloc byte[commandBytes.Length + 1] :
   new byte [commandBytes.Length + 1];

if (continuation)
{
   buffer[0] = 0x80;
}
else
{
   buffer[0] = 0;
}

commandBytes.CopyTo(buffer.Slice(1));

_i2cDevice.Write(buffer);
```

Reply:
Looking at the new behavior, something like this:
```C#
const int StackThreshold = 256;

byte[] commandBytes = command.GetBytes();

if (commandBytes == null) throw; // TODO: appropriate exception

if (commandBytes.IsEmpty) return; // TODO: Do nothing? throw exception? something else?

Span<byte> buffer = commandBytes.Length < StackThreshold ?
   stackalloc byte[commandBytes.Length + 1] :
   new byte [commandBytes.Length + 1]; // I don't think this will ever overflow int32 since we don't really have arrays with lengths == int.MaxValue

buffer[0] = 0; // may or may not be necessary if stack is zero-init'd. Keep it for now

commandBytes.CopyTo(buffer.Slice(1));

_i2cDevice.Write(buffer);
```

Reply:
@jaredpar, @agocke - for assemblies outside of corefx, is stackalloc span initialized with 0?

I am wondering if we need to explicitly set the first byte to 0 or not.

```C#
Span<byte> buffer = commandBytes.Length < StackThreshold ?
   stackalloc byte[commandBytes.Length + 1] :
   new byte [commandBytes.Length + 1]; // I don't think this will ever overflow int32 since we don't really have arrays with lengths == int.MaxValue

buffer[0] = 0; // may or may not be necessary if stack is zero-init'd. Keep it for now
```

Reply:
> for assemblies outside of corefx, is stackalloc span initialized with 0?

yes.

Reply:
Great! We can skip the unnecessary zero-init of `buffer[0]` then.

Reply:
Data size could vary depending on screen and such.  I usually use something like 32 messages of ~16 bytes to clear screen.  I updated the code to something like below.  Is this in the likes of your feedback?  I used 512 for threshold.

```csharp
const int stackThreshold = 512;

if (data == null || data.Length == 0)
{
   throw new ArgumentNullException(nameof(data));
}

Span<byte> writeBuffer = data.Length < stackThreshold ?
   stackalloc byte[data.Length + 1] :
   new byte[data.Length + 1];

writeBuffer[0] = 0x40; // Control byte.
data.CopyTo(writeBuffer.Slice(1));
_i2cDevice.Write(writeBuffer);
```

Reply:
> Is this in the likes of your feedback? I used 512 for threshold.

If `data.Length == 0` is invalid, I wouldn't throw argument null exception for it. Maybe `InvalidOperationException` or `ArgumentException`? Alternatively, we could just ignore it and do nothing (i.e. only write `if data.Length > 0`, otherwise ignore the command).

```C#
if (data == null)
{
   throw new ArgumentNullException(nameof(data));
}
if (data.Length == 0)
{
   throw new InvalidOperationException(nameof(data), "Data to write must not be empty.");
}
```

Otherwise, LGTM.

Reply:
@ahsonkhan just saw your comments before adding my last. Let me review tomorrow.  Disregard some of my last changes.  Thx again for the help.

Reply:
It seems like you can send empty data with just byte that represents a data command, so I'm just going to check for nulls for now.